File: java-mvc/layered-architecture/src/main/java/com/ecommerce/layered/business/service/ProductService.java
package com.ecommerce.layered.business.service;

import com.ecommerce.layered.business.exception.DuplicateSkuException;
import com.ecommerce.layered.business.exception.InsufficientStockException;
import com.ecommerce.layered.business.exception.ProductNotFoundException;
import com.ecommerce.layered.data.entity.ProductEntity;
import com.ecommerce.layered.data.repository.ProductRepository;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.List;
import java.util.UUID;

/**
 * BUSINESS LAYER - Service class.
 * Contains business logic and validation rules.
 * Depends on DATA LAYER (ProductRepository).
 * The Presentation layer depends on this.
 *
 * In layered architecture, the service directly uses the data entity.
 * This is simpler but creates tighter coupling between layers.
 */
@Service
@Transactional
public class ProductService {

    private final ProductRepository productRepository;

    public ProductService(ProductRepository productRepository) {
        this.productRepository = productRepository;
    }

    public ProductEntity createProduct(String name, String description, BigDecimal price,
                                        String category, Integer stockQuantity, String sku) {
        // Business validation
        if (sku != null && productRepository.existsBySku(sku)) {
            throw new DuplicateSkuException(sku);
        }

        ProductEntity product = new ProductEntity();
        product.setName(name);
        product.setDescription(description);
        product.setPrice(price);
        product.setCategory(category);
        product.setStockQuantity(stockQuantity);
        product.setSku(sku);
        product.setActive(true);

        return productRepository.save(product);
    }

    @Transactional(readOnly = true)
    public ProductEntity getProductById(UUID id) {
        return productRepository.findById(id)
                .orElseThrow(() -> new ProductNotFoundException(id));
    }

    @Transactional(readOnly = true)
    public Page<ProductEntity> getAllProducts(int page, int size) {
        size = Math.min(size, 100); // Max page size
        PageRequest pageRequest = PageRequest.of(page, size, Sort.by("createdAt").descending());
        return productRepository.findByActiveTrue(pageRequest);
    }

    @Transactional(readOnly = true)
    public List<ProductEntity> searchProducts(String query, String category,
                                               BigDecimal minPrice, BigDecimal maxPrice) {
        return productRepository.search(query, category, minPrice, maxPrice);
    }

    public ProductEntity updateProduct(UUID id, String name, String description,
                                        BigDecimal price, String category) {
        ProductEntity product = productRepository.findById(id)
                .orElseThrow(() -> new ProductNotFoundException(id));

        product.setName(name);
        product.setDescription(description);
        product.setPrice(price);
        product.setCategory(category);

        return productRepository.save(product);
    }

    public void deleteProduct(UUID id) {
        ProductEntity product = productRepository.findById(id)
                .orElseThrow(() -> new ProductNotFoundException(id));
        product.setActive(false);
        productRepository.save(product);
    }

    public void decreaseStock(UUID productId, int quantity) {
        if (quantity <= 0) {
            throw new IllegalArgumentException("Quantity must be positive");
        }

        ProductEntity product = productRepository.findById(productId)
                .orElseThrow(() -> new ProductNotFoundException(productId));

        if (product.getStockQuantity() < quantity) {
            throw new InsufficientStockException(product.getStockQuantity(), quantity);
        }

        product.setStockQuantity(product.getStockQuantity() - quantity);
        productRepository.save(product);
    }

    public void increaseStock(UUID productId, int quantity) {
        if (quantity <= 0) {
            throw new IllegalArgumentException("Quantity must be positive");
        }

        ProductEntity product = productRepository.findById(productId)
                .orElseThrow(() -> new ProductNotFoundException(productId));

        product.setStockQuantity(product.getStockQuantity() + quantity);
        productRepository.save(product);
    }

    @Transactional(readOnly = true)
    public long countProducts() {
        return productRepository.count();
    }
}

File: dotnet-mvc/LayeredArchitecture/Business/Services/ProductService.cs
using LayeredArchitecture.Business.Exceptions;
using LayeredArchitecture.Data.Entities;
using LayeredArchitecture.Data.Repositories;

namespace LayeredArchitecture.Business.Services;

/// <summary>
/// Service containing business logic for Product operations.
/// This belongs to the BUSINESS LAYER. In a layered architecture, the service
/// directly depends on the Data layer (repository and entities).
/// Dependencies flow downward: Presentation → Business → Data.
/// </summary>
public class ProductService
{
    private readonly ProductRepository _repository;

    public ProductService(ProductRepository repository)
    {
        _repository = repository;
    }

    public async Task<ProductEntity> CreateProduct(string name, string description, decimal price, string category, int stockQuantity, string sku)
    {
        if (price <= 0)
            throw new BusinessRuleException("Price must be greater than zero.");

        if (string.IsNullOrWhiteSpace(name))
            throw new BusinessRuleException("Product name is required.");

        if (string.IsNullOrWhiteSpace(sku))
            throw new BusinessRuleException("SKU is required.");

        if (stockQuantity < 0)
            throw new BusinessRuleException("Stock quantity cannot be negative.");

        if (await _repository.ExistsBySku(sku))
            throw new DuplicateSkuException(sku);

        var entity = new ProductEntity
        {
            Id = Guid.NewGuid(),
            Name = name,
            Description = description ?? string.Empty,
            Price = price,
            Category = category ?? string.Empty,
            StockQuantity = stockQuantity,
            Sku = sku,
            Active = true,
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow
        };

        return await _repository.Save(entity);
    }

    public async Task<ProductEntity> GetProductById(Guid id)
    {
        var entity = await _repository.FindById(id);
        if (entity is null)
            throw new ProductNotFoundException(id);

        return entity;
    }

    public async Task<(List<ProductEntity> Items, int TotalCount)> ListProducts(int page, int pageSize)
    {
        if (pageSize > 100) pageSize = 100;
        if (pageSize < 1) pageSize = 10;
        if (page < 0) page = 0;

        return await _repository.FindAll(page, pageSize);
    }

    public async Task<List<ProductEntity>> SearchProducts(string query)
    {
        if (string.IsNullOrWhiteSpace(query))
            return new List<ProductEntity>();

        return await _repository.Search(query);
    }

    public async Task<ProductEntity> UpdateProduct(Guid id, string name, string description, decimal price, string category, string sku)
    {
        if (price <= 0)
            throw new BusinessRuleException("Price must be greater than zero.");

        if (string.IsNullOrWhiteSpace(name))
            throw new BusinessRuleException("Product name is required.");

        var entity = await _repository.FindById(id);
        if (entity is null)
            throw new ProductNotFoundException(id);

        if (await _repository.ExistsBySkuAndIdNot(sku, id))
            throw new DuplicateSkuException(sku);

        entity.Name = name;
        entity.Description = description ?? string.Empty;
        entity.Price = price;
        entity.Category = category ?? string.Empty;
        entity.Sku = sku;
        entity.UpdatedAt = DateTime.UtcNow;

        return await _repository.Update(entity);
    }

    public async Task DeleteProduct(Guid id)
    {
        var entity = await _repository.FindById(id);
        if (entity is null)
            throw new ProductNotFoundException(id);

        entity.Active = false;
        entity.UpdatedAt = DateTime.UtcNow;
        await _repository.Update(entity);
    }

    public async Task<ProductEntity> DecreaseStock(Guid id, int quantity)
    {
        if (quantity <= 0)
            throw new BusinessRuleException("Quantity must be greater than zero.");

        var entity = await _repository.FindById(id);
        if (entity is null)
            throw new ProductNotFoundException(id);

        if (entity.StockQuantity - quantity < 0)
            throw new InsufficientStockException();

        entity.StockQuantity -= quantity;
        entity.UpdatedAt = DateTime.UtcNow;
        return await _repository.Update(entity);
    }

    public async Task<ProductEntity> IncreaseStock(Guid id, int quantity)
    {
        if (quantity <= 0)
            throw new BusinessRuleException("Quantity must be greater than zero.");

        var entity = await _repository.FindById(id);
        if (entity is null)
            throw new ProductNotFoundException(id);

        entity.StockQuantity += quantity;
        entity.UpdatedAt = DateTime.UtcNow;
        return await _repository.Update(entity);
    }
}
